pkg/operator/controllers/azurensg/subnetnsg.go

```diff
@@ -0,0 +1,173 @@
+	}
+	if fullSubnet.SubnetPropertiesFormat.NetworkSecurityGroup == nil {
+		return fmt.Errorf("received nil NSG when trying to Get subnet %s/%s in resource group %s", vnet.ResourceName, subnetName, vnet.ResourceGroup)
+	}
+	correctNSGResourceID, err := subnet.NetworkSecurityGroupIDExpanded(architectureVersion, masterResourceGroup, subnetName, infraID, isWorkerSubnet)
+	correctNSGResourceID = "/subscriptions/"+ vnet.SubscriptionID +"/resourceGroups/" + correctNSGResourceID
+	if err != nil {
+		return err
+	}
+	if !strings.EqualFold(*fullSubnet.NetworkSecurityGroup.ID,correctNSGResourceID) {
+		//NSG doesn't match - fixing
+		fullSubnet.NetworkSecurityGroup = &mgmtnetwork.SecurityGroup{ID: &correctNSGResourceID}
+		err = subnetsClient.CreateOrUpdateAndWait(ctx, vnet.ResourceGroup, vnet.ResourceName, subnetName, fullSubnet)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *AzureNSGReconciler) workerSubnets(ctx context.Context, maocli maoclient.Interface) ([]string, error) {
+	subnetMap := make(map[string]bool)
+	machines, err := maocli.MachineV1beta1().Machines(machineSetsNamespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return []string{}, err
+	}
+	for _, machine := range machines.Items {
+		if isMaster, err := r.isMasterRole(&machine); err == nil && !isMaster {
+			o, _, err := scheme.Codecs.UniversalDeserializer().Decode(machine.Spec.ProviderSpec.Value.Raw, nil, nil)
+			if err != nil {
+				return []string{}, err
+			}
+
+			machineProviderSpec, ok := o.(*azureproviderv1beta1.AzureMachineProviderSpec)
+			if !ok {
+				return []string{}, fmt.Errorf("machine %s: failed to read provider spec: %T", machine.Name, o)
+			}
+			subnetMap[machineProviderSpec.Subnet] = true
+		}
+	}
+	subnets := make([]string, len(subnetMap))
+
+	i := 0
+	for subnet := range subnetMap {
+		subnets[i] = subnet
+		i++
+	}
+	return subnets, nil
+}
+
+func (r *AzureNSGReconciler) masterParameters(ctx context.Context, maocli maoclient.Interface, vnetID string) (string, string, error) {
+	var masterSubnet, masterResourceGroup string
+	machines, err := maocli.MachineV1beta1().Machines(machineSetsNamespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return masterSubnet, masterResourceGroup, err
+	}
+	for _, machine := range machines.Items {
+		if isMaster, err := r.isMasterRole(&machine); err == nil && isMaster {
+			o, _, err := scheme.Codecs.UniversalDeserializer().Decode(machine.Spec.ProviderSpec.Value.Raw, nil, nil)
+			if err != nil {
+				return masterSubnet, masterResourceGroup, err
+			}
+
+			machineProviderSpec, ok := o.(*azureproviderv1beta1.AzureMachineProviderSpec)
+			if !ok {
+				return masterSubnet, masterResourceGroup, fmt.Errorf("machine %s: failed to read provider spec: %T", machine.Name, o)
+			}
+			masterSubnet = machineProviderSpec.Subnet
+			masterResourceGroup = machineProviderSpec.ResourceGroup
```

Comment:
This function bit stinks. We need to fix it. At the same time `workerSubnets` too.
I think this way:
* Add `clusterResourceGroup` into Operator object. It was long missing there.
* Refactor this function to be more user friendly to return list of subnets in all cluster (worker can be on multiple subnets). Something like `map[string]bool` where `map[subnetID]isMaster`. Function should have its own dedicated unit test. 
* And based on this we can do the logic in bit nicer way. And this second function should be unit testable too as it will just get input of the map

WDYT?
cc @m1kola